polynote-server/src/main/scala/polynote/server/repository/NotebookRepository.scala

```diff
@@ -225,16 +349,208 @@ class FileBasedRepository(
+        case (ver, notebook) => ver -> notebook.updateCell(cellID)(result.toCellUpdate)
+      }
+    } <* setNeedSave
+
+    override def clearResults(cellID: CellID): IO[AlreadyClosed, Unit] = ifOpen {
+      current.update {
+        case (ver, notebook) => ver -> notebook.updateCell(cellID)(_.copy(results = ShortList.Nil))
+      }
+    } <* setNeedSave
+
+    override def clearAllResults(): IO[AlreadyClosed, List[CellID]] = ifOpen {
+      current.modify {
+        case (ver, notebook) =>
+          val (clearedIds, updatedCells) = notebook.cells.foldRight((List.empty[CellID], List.empty[NotebookCell])) {
+            case (cell, (clearedIds, updatedCells)) =>
+              if (cell.results.nonEmpty)
+                (cell.id :: clearedIds) -> (cell.copy(results = ShortList(Nil)) :: updatedCells)
+              else
+                clearedIds -> (cell :: updatedCells)
+          }
+          clearedIds -> (ver -> notebook.copy(cells = ShortList(updatedCells)))
+      }
+    } <* setNeedSave
+
+    override def rename(newPath: String): RIO[BaseEnv with GlobalEnv, String] = renameLock.withPermit {
+      for {
+        currentPath <- path
+        newPath     <- renameNotebook(currentPath, newPath)
+        _           <- current.update {
+          case (ver, notebook) => ver -> notebook.copy(path = ShortString(newPath))
+        }
+      } yield newPath
+    } <* setNeedSave
+
+    override def close(): Task[Unit] =
+      closed.succeed(()) *>
+        pending.offer(Take.End) *>
+        pending.awaitShutdown *>
+        process.await.flatMap(_.join) *>
+        closed.await
+
+    override val isOpen: UIO[Boolean] = closed.isDone.map(!_)
+
+    override def awaitClosed: Task[Unit] = closed.await.ensuring(process.await.flatMap(_.await.unit))
+
+    private val verify = get.flatMap {
+      notebook =>
+        loadNotebook(notebook.path).flatMap {
+          loaded =>
+            val verified = loaded.cells.zip(notebook.cells).forall {
+              case (a, b) => a.content.toString == b.content.toString
+            }
+            if (verified) ZIO.unit else ZIO.fail(new IllegalStateException("Persisted notebook doesn't match in-memory notebook"))
+        }.catchAll {
+          err => encodeNotebook(notebook).flatMap {
+            encoded =>
+              Logging.error(
+                s"""Notebook verification failed for ${notebook.path} – were there filesystem errors during reading or writing?
+                   |Dumping to console:
+                   |$encoded
+                   |""".stripMargin,
+                err
+              )
+          }.orDie
+        }
+    }
+
+    private def init(): URIO[BaseEnv with GlobalEnv, Unit] = {
+      def doUpdate(update: NotebookUpdate) = current.updateAndGet {
+        case (prevVer, notebook) =>
+          val nextVer = prevVer + 1
+          val next = try {
+            // TODO: Notebook updates should be total, but they're currently not because InsertCell is assigned
```

Comment:
Agreed, we should have the server mint all IDs I think. Some of the work I'm doing in the UI refactor should help here.